x-pack/plugins/security_solution/server/endpoint/services/actions/clients/crowdstrike/crowdstrike_actions_client.ts

```diff
@@ -0,0 +1,341 @@
+
+export type CrowdstrikeActionsClientOptions = ResponseActionsClientOptions & {
+  connectorActions: NormalizedExternalConnectorClient;
+};
+
+export class CrowdstrikeActionsClient extends ResponseActionsClientImpl {
+  protected readonly agentType: ResponseActionAgentType = 'crowdstrike';
+  private readonly connectorActionsClient: NormalizedExternalConnectorClient;
+
+  constructor({ connectorActions, ...options }: CrowdstrikeActionsClientOptions) {
+    super(options);
+    this.connectorActionsClient = connectorActions;
+    connectorActions.setup(CROWDSTRIKE_CONNECTOR_ID);
+  }
+
+  protected async writeActionRequestToEndpointIndex<
+    TParameters extends EndpointActionDataParameterTypes = EndpointActionDataParameterTypes,
+    TOutputContent extends EndpointActionResponseDataOutput = EndpointActionResponseDataOutput,
+    TMeta extends {} = {}
+  >(
+    actionRequest: ResponseActionsClientWriteActionRequestToEndpointIndexOptions<
+      TParameters,
+      TOutputContent,
+      TMeta
+    >
+  ): Promise<
+    LogsEndpointAction<TParameters, TOutputContent, TMeta & CrowdstrikeActionRequestCommonMeta>
+  > {
+    const agentId = actionRequest.endpoint_ids[0];
+    const eventDetails = await this.getEventDetailsById(agentId);
+
+    const hostname = eventDetails.crowdstrike.event.HostName;
+    return super.writeActionRequestToEndpointIndex({
+      ...actionRequest,
+      hosts: {
+        [agentId]: { name: hostname },
+      },
+      meta: {
+        hostName: hostname,
+        ...(actionRequest.meta ?? {}),
+      } as TMeta & CrowdstrikeActionRequestCommonMeta,
+    });
+  }
+
+  /**
+   * Sends actions to Crowdstrike directly (via Connector)
+   * @private
+   */
+  private async sendAction(
+    actionType: SUB_ACTION,
+    actionParams: object
+  ): Promise<ActionTypeExecutorResult<unknown>> {
+    const executeOptions: NormalizedExternalConnectorClientExecuteOptions = {
+      params: {
+        subAction: actionType,
+        subActionParams: actionParams,
+      },
+    };
+
+    this.log.debug(
+      `calling connector actions 'execute()' for Crowdstrike with:\n${stringify(executeOptions)}`
+    );
+
+    const actionSendResponse = await this.connectorActionsClient.execute(executeOptions);
+
+    if (actionSendResponse.status === 'error') {
+      this.log.error(stringify(actionSendResponse));
+
+      throw new ResponseActionsClientError(
+        `Attempt to send [${actionType}] to Crowdstrike failed: ${
+          actionSendResponse.serviceMessage || actionSendResponse.message
+        }`,
+        500,
+        actionSendResponse
+      );
+    } else {
+      this.log.debug(`Response:\n${stringify(actionSendResponse)}`);
+    }
+
+    return actionSendResponse;
+  }
+
+  private async getEventDetailsById(agentId: string): Promise<{
+    crowdstrike: { event: { HostName: string } };
+  }> {
+    const search = {
+      index: ['logs-crowdstrike.fdr*', 'logs-crowdstrike.falcon*'],
+      fields: [{ field: '*' }],
```

Comment:
I'm curious why you used `fields` and not just a regular query and get the values from `_source`. can you explain?

Reply:
At first I used a filter here just to get 1 value, but then I remembered what you said once to gather all the data, but that actually might have been for a agent details usage. 
Is it fine for you if I just query for `fields: [{ field: 'crowdstrike.event.HostName' }],` ?

Reply:
yeah, if that works and thats the only value you need sure. 

but whats the reasoning for using `fields` vs just accessing the data you need via `_source`?

`fields` seems to flatten the property access path in the response (just tried it in the console), but then on line 143 below you seem to be converting itt back to a nested object, so I would like to understand this a bit more

Reply:
You're right, I used fields just for a reason of getting a single field, not all object. 
But also: sometime ago there were requests from security_solution team asking to remove _source because it's 'deprecated'. I am trying to avoid using _source since then. But can't really find good arguments WHY source would be bad.

Reply:
OHhh. .thats news to me (if you can point me to where you found that info, that would be great). I have not heard that and have seen nothing from kibana core on this. They would normally `deprecate` the `_source` property so that devs get a visual to stop using it (just like they recently did for `body` attribute of the search options)

I'm wondering if what you are really looking for is something like this:

```
{
  // ...

  _source: [ 'crowdstrike.event.HostName' ]
}
```

Anyway I'll leave it to you to decide what implementation to keep here. I was just wanted to better understand the reasoning. 

_(Note: I would still like to know about why two index patterns are being used - see my comment for line 127)_

Reply:
I'll try tomorrow with the  _source: [ 'crowdstrike.event.HostName' ] , maybe then I would be able to skip using extendObject and typing would become easier. 👍

Reply:
_source: [ 'crowdstrike.event.HostName' ] worked nicely, thanks for the hint, I didn't know we could specify fields through source.